cupy/_core/_routines_linalg.pyx

```diff
@@ -726,7 +726,15 @@ cpdef ndarray matmul(ndarray a, ndarray b, ndarray out=None):
 
     ndim = max(orig_a_ndim, orig_b_ndim)
     if ndim <= 2:
-        return dot(a, b, out)
+        if out is None or out._c_contiguous:
```

Comment:
```suggestion
        if out is None:
```
because the condition `out._c_contiguous and ret_dtype == out.dtype` below is correct for `out is not None` cases.